keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+This KEP proposes to initiate pod termination on the status of application containers. Usually a pod with multiple containers have core application containers and some sidecar containers like logging, service mesh. We are proposing to distinguish between application containers and sidecar containers using a pod annotation and initiating pod termination  if application containers are terminated.
+
+This is particularly helpful in solving the blocking issue of sidecar containers with jobs. This will also help us to make a step forward with the sidecar KEP. 
+
+We are calling `application/essential/main/critical` containers as `keystone` containers in this KEP to distnguish it from sidecars.
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+There are lots of [user requests](https://https://github.com/kubernetes/kubernetes/issues/25908) for better support of sidecar containers in Jobs but this issue mostly gets lost in a larger attempt to address the sidecar issue as a whole.
+
+Consider a K8s job with two containers one that does the actual work and another one is for monitoring/logging, the keystone container will complete the work and exit, the sidecar container will keep running and the job will never run to completion. This is problematic for any external application/tool depending on the job waiting to gets completed. Also, the sidecars unnecessarily occupy the system resources.
```

Comment:
What about when the sidecar crashes? Won't that cause an unnecessary restart of the application? Is that in scope?

Reply:
For the moment this is out of scope. If sidecar container crashes it might be restarted, but won't cause the pod to fail or restart.

Reply:
the usual restart policy will be applied if the sidecar container crashes, currently there will be no change in pod lifecycle if sidecar container crashes , added this in `Non-Goals`.

as @matthyx said if it is much required we can add it